javascript/ql/lib/semmle/javascript/endpoints/EndpointNaming.qll

```diff
@@ -0,0 +1,475 @@
+  shortestDistances(isPackageExport/1, memberEdge/2)(_, nd, result)
+
+private predicate isExported(API::Node node) {
+  isPackageExport(node)
+  or
+  exists(API::Node pred |
+    isExported(pred) and
+    memberEdge(pred, node)
+  )
+}
+
+/**
+ * Holds if `node` is a default export that can be reinterpreted as a namespace export,
+ * because the enclosing module has no named exports.
+ */
+private predicate defaultExportCanBeInterpretedAsNamespaceExport(API::Node node) {
+  exists(ES2015Module mod |
+    node.asSink() = mod.getAnExportedValue("default") and
+    not mod.hasBothNamedAndDefaultExports()
+  )
+}
+
+private predicate isPrivateAssignment(DataFlow::Node node) {
+  exists(MemberDeclaration decl |
+    node = decl.getInit().flow() and
+    decl.isPrivate()
+  )
+  or
+  exists(DataFlow::PropWrite write |
+    write.isPrivateField() and
+    node = write.getRhs()
+  )
+}
+
+private predicate isPrivateLike(API::Node node) { isPrivateAssignment(node.asSink()) }
+
+private API::Node getASuccessor(API::Node node, string name, int badness) {
+  isExported(node) and
+  exists(string member |
+    result = node.getMember(member) and
+    not isPrivateLike(node) and
+    if member = "default"
+    then
+      if defaultExportCanBeInterpretedAsNamespaceExport(node)
+      then (
+        badness = 5 and name = ""
+      ) else (
+        badness = 10 and name = "default"
+      )
+    else (
+      name = member and badness = 0
+    )
+  )
+}
+
+private API::Node getAPredecessor(API::Node node, string name, int badness) {
+  node = getASuccessor(result, name, badness)
+}
+
+/**
+ * Gets the predecessor of `node` to use when constructing a qualified name for it,
+ * and binds `name` and `badness` corresponding to the label on that edge.
+ */
+private API::Node getPreferredPredecessor(API::Node node, string name, int badness) {
+  // For root nodes, we prefer not having a predecessor, as we use the package name.
+  not isPackageExport(node) and
+  // Rank predecessors by name-badness, export-distance, and name.
+  // Since min() can only return a single value, we need a separate min() call per column.
+  badness = min(int b | exists(getAPredecessor(node, _, b)) | b) and
+  result =
+    min(API::Node pred, string name1 |
+      pred = getAPredecessor(node, name1, badness)
+    |
+      pred order by distanceFromPackageExport(pred), name1
+    ) and
+  name = min(string n | result = getAPredecessor(node, n, badness) | n)
+}
+
+/**
+ * Holds if values escpin
+ */
+private predicate sinkHasNameCandidate(API::Node sink, string package, string name, int badness) {
+  sink = API::moduleExport(package) and
+  name = "" and
+  badness = 0
+  or
+  exists(API::Node baseNode, string baseName, int baseBadness, string step, int stepBadness |
+    sinkHasNameCandidate(baseNode, package, baseName, baseBadness) and
+    baseNode = getPreferredPredecessor(sink, step, stepBadness) and
+    badness = (baseBadness + stepBadness).minimum(20) and
```

Comment:
I don't get this 20.

Reply:
It was needed to ensure termination at some point, but is not needed anymore. I've fixed the criteria in `getPreferredPredecessor` to completely rule out the possibility of a cycle of preferred predecessor edges.